engine/access/rest/websockets/controller.go

```diff
@@ -111,214 +113,277 @@ func (c *Controller) configureKeepalive() error {
 			}
-			// TODO: handle 'response per second' limits
 
-			// Specifies a timeout for the write operation. If the write
-			// isn't completed within this duration, it fails with a timeout error.
-			// SetWriteDeadline ensures the write operation does not block indefinitely
-			// if the client is slow or unresponsive. This prevents resource exhaustion
-			// and allows the server to gracefully handle timeouts for delayed writes.
 			if err := c.conn.SetWriteDeadline(time.Now().Add(WriteWait)); err != nil {
 				return fmt.Errorf("failed to set the write deadline: %w", err)
 			}
-			err := c.conn.WriteJSON(msg)
-			if err != nil {
-				return fmt.Errorf("failed to write message to connection: %w", err)
+
+			if err := c.conn.WriteJSON(message); err != nil {
+				return err
 			}
 		}
 	}
 }
 
 // readMessages continuously reads messages from a client WebSocket connection,
-// processes each message, and handles actions based on the message type.
+// validates each message, and processes it based on the message type.
 //
-// No errors are expected during normal operation. All errors are considered benign.
+// Expected errors during normal operation:
+// - context.Canceled if the client disconnected
 func (c *Controller) readMessages(ctx context.Context) error {
 	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		default:
-			msg, err := c.readMessage()
-			if err != nil {
-				if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseAbnormalClosure) {
-					return nil
-				}
-				return fmt.Errorf("failed to read message from client: %w", err)
+		var message json.RawMessage
+		if err := c.conn.ReadJSON(&message); err != nil {
+			if errors.Is(err, websocket.ErrCloseSent) {
+				return err
 			}
 
-			_, validatedMsg, err := c.parseAndValidateMessage(msg)
-			if err != nil {
-				//TODO: write error to error channel
-				return fmt.Errorf("failed to parse and validate client message: %w", err)
-			}
-
-			if err := c.handleAction(ctx, validatedMsg); err != nil {
-				//TODO: write error to error channel
-				return fmt.Errorf("failed to handle message action: %w", err)
-			}
+			c.writeErrorResponse(
+				ctx,
+				err,
+				wrapErrorMessage(InvalidMessage, "error reading message", "", "", ""))
+			continue
 		}
-	}
-}
 
-func (c *Controller) readMessage() (json.RawMessage, error) {
-	var message json.RawMessage
-	if err := c.conn.ReadJSON(&message); err != nil {
-		return nil, fmt.Errorf("error reading JSON from client: %w", err)
+		err := c.parseAndValidateMessage(ctx, message)
+		if err != nil {
+			c.writeErrorResponse(
+				ctx,
+				err,
+				wrapErrorMessage(InvalidMessage, "error parsing message", "", "", ""))
+			continue
+		}
 	}
-	return message, nil
 }
 
-func (c *Controller) parseAndValidateMessage(message json.RawMessage) (models.BaseMessageRequest, interface{}, error) {
+func (c *Controller) parseAndValidateMessage(ctx context.Context, message json.RawMessage) error {
```

Comment:
this does more than `parseAndValidateMessage`. how about renaming to `handleMessage`?